webapp/web/routes_test.go

```diff
@@ -121,6 +121,27 @@ func TestAdminCacheFlushBound(t *testing.T) {
 	assertHSTS(t, "/admin/cache/flush")
 }
 
+func TestApiMetadataCORS(t *testing.T) {
+	assertHandlerIs(t, "/api/metadata", "api-metadata")
+
+	successReq := httptest.NewRequest("OPTIONS", "/api/metadata", nil)
+	successReq.Header.Set("Access-Control-Request-Headers", "content-type")
+	successReq.Header.Add("Origin", "https://developer.mozilla.org")
+	successReq.Header.Add("Access-Control-Request-Method", "PATCH")
+	rr := sendHttptestRequest(successReq)
+
+	assert.Equal(t, http.StatusOK, rr.Result().StatusCode)
+	assert.Equal(t, "Content-Type", rr.Result().Header.Get("Access-Control-Allow-Headers"))
+	assert.Equal(t, "PATCH", rr.Result().Header.Get("Access-Control-Allow-Methods"))
+	assert.Equal(t, "https://developer.mozilla.org", rr.Result().Header.Get("Access-Control-Allow-Origin"))
+
+	invalidOriginReq := httptest.NewRequest("PATCH", "/api/metadata", nil)
+	invalidOriginReq.Header.Add("Origin", "https://foo")
+	rr = sendHttptestRequest(invalidOriginReq)
+
+	assert.Equal(t, "", rr.Result().Header.Get("Access-Control-Allow-Origin"))
```

Comment:
We want to test the following cases.

* For the read-only/permissive API
    * A simple GET request from an untrusted origin: the response should contain `Access-Control-Allow-Origin: *` and `Access-Control-Allow-Headers: Content-Type`.
    * A preflight OPTIONS request for POST with `Access-Control-Request-Headers: Content-Type` from an untrusted origin: the response should contain `Access-Control-Allow-Origin: *` and `Access-Control-Allow-Headers: Content-Type`.
    * A POST request with `Content-Type: application/json` from an untrusted origin: the response should contain `Access-Control-Allow-Origin: *`.
* For the mutation/restrictive API
    * A preflight OPTIONS request for PATCH from an untrusted origin: the response should **not** contain `Access-Control-Allow-Origin`.
    * A PATCH request from an untrusted origin: the response should **not** contain `Access-Control-Allow-Origin`.
    * A preflight OPTIONS request for PATCH with `Access-Control-Request-Headers: Content-Type` from a trusted origin: the response should contain `Access-Control-Allow-Origin` matching the origin, `Access-Control-Allow-Headers: Content-Type`, and `Access-Control-Allow-Credentials`.
    * A PATCH request from a trusted origin: the response should contain `Access-Control-Allow-Origin` matching the origin.

Reply:
Thanks for the clarification. I didn't understand the CORS correctly for simple requests when we chatted offline. Due to the fact that these test requests does not use browsers (it doesn't trigger preflights automatically and gorilla doesn't response to any non-OPTIONS methods), it would be hard to test GET/POST/PATCH requests, without creating mocks for loggers, httpClients.

Reply:
Let's leave a TODO here for now.